### PR TITLE
[5.8] Resubmission of PR #27325 - Fix numeric keys in resources

### PR DESCRIPTION
**Pull request by rodrigopedra**
This PR is a resubmission of PR #27325 . It adds the ability to preserve the keys of an associative array if it has custom indices.

Currently any array returned in a `Resource` will be converted to a plain JSON array if its first key is numeric.

So assuming you have this Resource:

```php
<?php

namespace App\Http\Resources;

use Illuminate\Http\Resources\Json\JsonResource;

class MyResource extends JsonResource
{
    public function toArray( $request )
    {
        return [
            'data' => $this->resource[ 'data' ],
        ];
    }
}
```

And try to return this resource using the data below:

```php
$router->get('test', function (  ) {
    $data = [ 'data'=> [ '1' => 10, '2' => 20, 'total' => 30 ] ];

    return new \App\Http\Resources\MyResource($data);
});
```

As of 5.7, as the first index is numeric (`1`) the `JsonResource` class would strips all the index and would return only the values as an array:

```json
{"data":[10,20,30]}
```

This PR, adds the ability to define a `$preserveKeys` property in your Resource, so when the JSON output is Serialized the keys are preserved. For example:

```php
<?php

namespace App\Http\Resources;

use Illuminate\Http\Resources\Json\JsonResource;

class MyResource extends JsonResource
{
    public $preserveKeys = true; // flag to preserve keys during JSON serialization

    public function toArray( $request )
    {
        return [
            'data' => $this->resource[ 'data' ],
        ];
    }
}
```
Using the same route as above, the output would be:

```json
{"data":{"1":10,"2":20,"total":30}}
```

If the `$preserveKeys` property is not set, or if it is set to `false` the current behavior is preserved.

Note that if the data is a "plain" array (all indices are sequential and start at zero, i.e. `array_values($arr) === $arr`) , a plain JSON array is returned even when `$preserveKeys = true`

---

Most of the code in this PR is based on previous work from @staudenmeir  in PR's #27325 and #24339

### CODE REVIEW

File: src/Illuminate/Http/Resources/ConditionallyLoadsAttributes.php [hunk 1]
```diff
@@ -28,15 +26,30 @@ protected function filter($data)
             }
 
             if (is_numeric($key) && $value instanceof MergeValue) {
-                return $this->mergeData($data, $index, $this->filter($value->data), $numericKeys);
+                return $this->mergeData($data, $index, $this->filter($value->data));
```
Comment (staudenmeir) [inline src/Illuminate/Http/Resources/ConditionallyLoadsAttributes.php]: This behaves differently than #27325:
#27325 determines `$numericKeys` on `$value->data`, your PR on `$this->filter($value->data)`.

You can see the incorrect result by adjusting the data in `test_leading_merge__keyed_value_is_merged_correctly`:

```php
new MergeValue([0 => new MissingValue, 'name' => 'mohamed', 'location' => 'hurghada'])
```

Comment (rodrigopedra) [reply src/Illuminate/Http/Resources/ConditionallyLoadsAttributes.php]: Thanks, I'll take a look tomorrow.

**Comment by rodrigopedra** [comment]
After thinking about the use cases described in PR's #27235 and #24339 I'll submit a new PR with a different approach

**Comment by rodrigopedra** [comment]
Subimitted new PR #27394 with a new simplified approach

**Pull request closed without merging**